File: src/routers/funding_rates.py
"""
Funding Rates Router for the AR trading application.

This module provides API endpoints for fetching funding rates
from multiple exchanges.
"""

from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Optional
from sqlalchemy.orm import Session

from src.database.session import get_db
from src.services.funding_rate_service import FundingRateService
from src.schemas.funding_rate_schema import (
    FundingRateResponse,
    FundingRatesSummaryResponse,
    ExchangeFundingRatesRequest,
    FundingRateItem
)
from src.utils.error_handlers import exchange_error_handler

router = APIRouter()

# Initialize funding rate service
funding_rate_service = FundingRateService()

@router.get("/funding-rates/", response_model=FundingRatesSummaryResponse)
@exchange_error_handler("funding_rates", "funding rates fetching")
async def get_all_funding_rates(
    symbols: Optional[List[str]] = Query(None, description="Specific symbols to fetch"),
    exchanges: Optional[List[str]] = Query(None, description="Specific exchanges to fetch from")
):
    """
    Get funding rates for all supported symbols across all supported exchanges.
    
    Args:
            symbols: Optional list of specific symbols (e.g., ['BTC', 'ETH'])
    exchanges: Optional list of specific exchanges (e.g., ['kucoin', 'bitunix'])
        
    Returns:
        Complete funding rates summary for all symbols and exchanges
    """
    try:
        rates = await funding_rate_service.get_all_funding_rates(
            symbols=symbols,
            exchanges=exchanges
        )
        return rates
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch funding rates: {str(e)}"
        )

@router.get("/funding-rates/{symbol}", response_model=FundingRateResponse)
@exchange_error_handler("funding_rates", "symbol funding rates fetching")
async def get_symbol_funding_rates(
    symbol: str,
    exchanges: Optional[List[str]] = Query(None, description="Specific exchanges to fetch from")
):
    """
    Get funding rates for a specific symbol across all supported exchanges.
    
    Args:
        symbol: Base symbol (e.g., 'BTC', 'ETH')
        exchanges: Optional list of specific exchanges
        
    Returns:
        Funding rates for the specified symbol across exchanges
    """
    try:
        rates = await funding_rate_service.get_symbol_funding_rates(
            symbol=symbol.upper(),
            exchanges=exchanges
        )
        return rates
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch funding rates for {symbol}: {str(e)}"
        )

@router.get("/funding-rates/exchange/{exchange_id}", response_model=List[FundingRateItem])
@exchange_error_handler("funding_rates", "exchange funding rates fetching")
async def get_exchange_funding_rates(
    exchange_id: str,
    symbols: Optional[List[str]] = Query(None, description="Specific symbols to fetch")
):
    """
    Get funding rates for all symbols from a specific exchange.
    
    Args:
        exchange_id: Exchange identifier (e.g., 'kucoin', 'bitunix')
        symbols: Optional list of specific symbols
        
    Returns:
        List of funding rates for the specified exchange
    """
    try:
        rates = await funding_rate_service.get_exchange_funding_rates(
            exchange_id=exchange_id.lower(),
            symbols=symbols
        )
        return rates
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch funding rates from {exchange_id}: {str(e)}"
        )

@router.post("/funding-rates/custom", response_model=FundingRatesSummaryResponse)
@exchange_error_handler("funding_rates", "custom funding rates fetching")
async def get_custom_funding_rates(
    request: ExchangeFundingRatesRequest
):
    """
    Get funding rates with custom parameters.
    
    Args:
        request: Custom funding rates request with exchanges and symbols
        
    Returns:
        Funding rates summary based on custom parameters
    """
    try:
        rates = await funding_rate_service.get_all_funding_rates(
            symbols=request.symbols,
            exchanges=request.exchanges
        )
        return rates
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch custom funding rates: {str(e)}"
        )

@router.get("/funding-rates/supported/exchanges")
async def get_supported_exchanges():
    """
    Get list of supported exchanges for funding rates.
    
    Returns:
        List of supported exchange identifiers
    """
    from src.services.funding_rate_service import SUPPORTED_EXCHANGES
    return {
        "exchanges": SUPPORTED_EXCHANGES,
        "message": "Supported exchanges for funding rates"
    }

@router.get("/funding-rates/supported/symbols")
async def get_supported_symbols():
    """
    Get list of supported symbols for funding rates.
    
    Returns:
        List of supported symbols and their trading pairs
    """
    from src.services.funding_rate_service import SUPPORTED_SYMBOLS, SYMBOL_TO_PAIR
    return {
        "symbols": SUPPORTED_SYMBOLS,
        "trading_pairs": SYMBOL_TO_PAIR,
        "message": "Supported symbols and trading pairs for funding rates"
    }

@router.on_event("shutdown")
async def shutdown_event():
    """Clean up funding rate service on shutdown."""
    await funding_rate_service.cleanup()

File: frontend/src/api/funding_rates.js
/**
 * Funding Rates API helper functions.
 * 
 * This module provides functions to fetch funding rate data
 * from multiple exchanges.
 */

/**
 * Get all funding rates for all supported symbols and exchanges
 * @param {string[]} symbols - Optional array of symbols (e.g., ['BTC', 'ETH'])
 * @param {string[]} exchanges - Optional array of exchanges (e.g., ['kucoin', 'mexc'])
 * @returns {Promise<Object|null>} Funding rates data or null if error
 */
export const getAllFundingRates = async (symbols = null, exchanges = null) => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    let url = `${apiBaseUrl}/funding-rates/`;
    const params = new URLSearchParams();
    
    if (symbols && symbols.length > 0) {
      symbols.forEach(symbol => params.append('symbols', symbol));
    }
    
    if (exchanges && exchanges.length > 0) {
      exchanges.forEach(exchange => params.append('exchanges', exchange));
    }
    
    if (params.toString()) {
      url += `?${params.toString()}`;
    }
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {};
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, { headers });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error('Error fetching all funding rates:', response.statusText);
      return null;
    }
  } catch (error) {
    console.error('Error fetching all funding rates:', error);
    return null;
  }
};

/**
 * Get funding rates for a specific symbol across all exchanges
 * @param {string} symbol - Trading symbol (e.g., 'BTC')
 * @param {string[]} exchanges - Optional array of exchanges
 * @returns {Promise<Object|null>} Funding rates for the symbol or null if error
 */
export const getSymbolFundingRates = async (symbol, exchanges = null) => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    let url = `${apiBaseUrl}/funding-rates/${symbol}`;
    
    if (exchanges && exchanges.length > 0) {
      const params = new URLSearchParams();
      exchanges.forEach(exchange => params.append('exchanges', exchange));
      url += `?${params.toString()}`;
    }
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {};
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, { headers });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error(`Error fetching funding rates for ${symbol}:`, response.statusText);
      return null;
    }
  } catch (error) {
    console.error(`Error fetching funding rates for ${symbol}:`, error);
    return null;
  }
};

/**
 * Get funding rates for all symbols from a specific exchange
 * @param {string} exchangeId - Exchange identifier (e.g., 'kucoin')
 * @param {string[]} symbols - Optional array of symbols
 * @returns {Promise<Array|null>} Funding rates from the exchange or null if error
 */
export const getExchangeFundingRates = async (exchangeId, symbols = null) => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    let url = `${apiBaseUrl}/funding-rates/exchange/${exchangeId}`;
    
    if (symbols && symbols.length > 0) {
      const params = new URLSearchParams();
      symbols.forEach(symbol => params.append('symbols', symbol));
      url += `?${params.toString()}`;
    }
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {};
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, { headers });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error(`Error fetching funding rates from ${exchangeId}:`, response.statusText);
      return null;
    }
  } catch (error) {
    console.error(`Error fetching funding rates from ${exchangeId}:`, error);
    return null;
  }
};

/**
 * Get custom funding rates with specific parameters
 * @param {Object} request - Request object with exchanges and symbols
 * @returns {Promise<Object|null>} Custom funding rates data or null if error
 */
export const getCustomFundingRates = async (request) => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    const url = `${apiBaseUrl}/funding-rates/custom`;
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {
      'Content-Type': 'application/json'
    };
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(request)
    });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error('Error fetching custom funding rates:', response.statusText);
      return null;
    }
  } catch (error) {
    console.error('Error fetching custom funding rates:', error);
    return null;
  }
};

/**
 * Get list of supported exchanges for funding rates
 * @returns {Promise<Object|null>} Supported exchanges or null if error
 */
export const getSupportedExchanges = async () => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    const url = `${apiBaseUrl}/funding-rates/supported/exchanges`;
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {};
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, { headers });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error('Error fetching supported exchanges:', response.statusText);
      return null;
    }
  } catch (error) {
    console.error('Error fetching supported exchanges:', error);
    return null;
  }
};

/**
 * Get list of supported symbols for funding rates
 * @returns {Promise<Object|null>} Supported symbols or null if error
 */
export const getSupportedSymbols = async () => {
  try {
    const apiBaseUrl = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
    const url = `${apiBaseUrl}/funding-rates/supported/symbols`;
    
    const apiKey = localStorage.getItem('apiKey');
    const headers = {};
    if (apiKey) {
      headers['X-API-Key'] = apiKey;
    }
    
    const response = await fetch(url, { headers });
    
    if (response.ok) {
      return await response.json();
    } else {
      console.error('Error fetching supported symbols:', response.statusText);
      return null;
    }
  } catch (error) {
    console.error('Error fetching supported symbols:', error);
    return null;
  }
};
